wd_demo.au3

```diff
@@ -1269,7 +1269,7 @@ Func SetupEdgeIEMode() ; this is for MS Edge IE Mode
 	; https://www.selenium.dev/documentation/ie_driver_server/#required-configuration
 	Local $sTimeStamp = @YEAR & '-' & @MON & '-' & @MDAY & '_' & @HOUR & @MIN & @SEC
 	_WD_Option('Driver', 'IEDriverServer.exe') ;
-	_WD_Option('Port', 5555)
+	_WD_Option('Port', _WD_GetFreePort(5555, 5600))
```

Comment:
should we add 
```autoit
_WD_Option('Port', _WD_GetFreePort(5555, 5600))
```
also for other browsers ?

Reply:
I don't think so. Technically, it isn't even needed for `iemode` because everything still runs fine, albeit with some minor logging glitches.

Reply:
IMO it is possible that `iemod` on some slower PC could hit the problem very often and will be unable to start driver connection.

Reply:
It has been my experience the IEDriverServer would load and run even when the target port is in  TIME_WAIT state. I can immediately run IEDriverServer after it has exited a prior run.

Does that match your experience?

Reply:
Please try with this modified:

```
Func _WD_GetFreePort($iMinPort = Default, $iMaxPort = Default)
	Local Const $sFuncName = "_WD_GetFreePort"
	Local $iResult = 0, $iErr = $_WD_ERROR_Success

	If $iMaxPort = Default Then $iMaxPort = ($iMinPort = Default) ? 65000 : $iMinPort
	If $iMinPort = Default Then $iMinPort = 64000
	Local $aPorts = __WinAPI_GetTcpTable()
	If @error Then
		$iErr = $_WD_ERROR_GeneralError
	Else
		For $iPort = $iMinPort To $iMaxPort
			_ArraySearch($aPorts, $iPort, Default, Default, Default, Default, Default, 3)
			If @error = 6 Then ; not found in list of used port
				$iResult = $iPort
				ExitLoop
			EndIf
		Next
		If $iPort > $iMaxPort Then $iErr = $_WD_ERROR_NotFound
	EndIf

	Return SetError(__WD_Error($sFuncName, $iErr, $iResult), 0, $iResult)
EndFunc   ;==>_WD_GetFreePort
```

and:
```autoit
	_WD_Option('Driver', 'IEDriverServer.exe') ;
	_WD_Option('Port', _WD_GetFreePort(5555))
```

note:
I think that you hit an error from: `__WinAPI_GetTcpTable()`
IMO we should get two different errors from: _WD_GetFreePort 
$_WD_ERROR_GeneralError 
or 
$_WD_ERROR_NotFound

Reply:
> It has been my experience the IEDriverServer would load and run even when the target port is in TIME_WAIT state. I can immediately run IEDriverServer after it has exited a prior run.


I have used:
```autoit
	Local $aPorts = __WinAPI_GetTcpTable()
	_ArrayDisplay($aPorts)
```
and tested
```
C:\Windows\System32>netstat -aon | find "5555"
  TCP    127.0.0.1:5555         127.0.0.1:53183        TIME_WAIT       0
  TCP    127.0.0.1:5555         127.0.0.1:53252        TIME_WAIT       0
```
and `netstat` shows what you see above and `_ArrayDisplay` is showing this port as used.
![image](https://user-images.githubusercontent.com/11089482/233476762-2ba22399-52f6-4b5a-b1ff-88076500be7c.png)



> Does that match your experience?

No as I had no issue with: `_WD_Startup: Port: 5555 (Unavailable)`
which as I stated comes from: `__WinAPI_GetTcpTable()`

Reply:
> No as I had no issue with: _WD_Startup: Port: 5555 (Unavailable)

Run wd_demo, run multiple passes of `DemoNavigation`, then check the Scite output panel. Obviously, you would need to revert to `_WD_Option('Port', 5555)` for the test to be valid.

Reply:
> It has been my experience the IEDriverServer would load and run even when the target port is in TIME_WAIT state. I can immediately run IEDriverServer after it has exited a prior run.
> 
> Does that match your experience?

yes. I confirm

Reply:
> 	_WD_Option('Port', _WD_GetFreePort(5555))

There's no point in calling `_WD_GetFreePort` here (when a single port is provided) as it is automatically called in `_WD_Startup`. Doing it this way will assign 0 as the port number if the port is in use.

Reply:
oops 
```autoit
_WD_Option('Port', _WD_GetFreePort(5555, 5600))
```

Reply:
> and netstat shows what you see above and _ArrayDisplay is showing this port as used.

Right... which would explain why I received the "Unavailable" message because `_WD_GetFreePort` detected that the port was open. I agree with adding the additional error code. I'll submit a separate PR for this and other possible _WD_GetFreePort enhancements.

Reply:
or even better
```autoit
_WD_Option('Port', _WD_GetFreePort())
```

Reply:
> _WD_Option('Port', _WD_GetFreePort(5555, 5600))

I'm thinking that these should be separated into two statements, so that you could handle any errors. Otherwise, the error gets ignored and the port will equal 0.

Something like this --

```
Local $iPort = _WD_GetFreePort(5555, 5600))
If @error Then $iPort = 5555
_WD_Option('Port', $iPort)
```

Reply:
Here is better idea:

[Execute browser setup - error handling](https://github.com/Danp2/au3WebDriver/pull/454/commits/aa592a1f12ac8e4568a5e99a868684a00c81fbb2)

[SetupEdgeIEMode + _WD_GetFreePort](https://github.com/Danp2/au3WebDriver/pull/454/commits/aa458ff0d3725f2fd44750eff0a0d2a3f6c16cc6)

test it with:
```autoit
Local $iPort = _WD_GetFreePort(5555, 5555)
```

Reply:
FWIW, I like my way better because it doesn't abort due to an issue with `__WinAPI_GetTcpTable`. It also should be easier for the user to determine what went wrong, whereas your version simply exits without any type of explanation AFAICS.

Reply:
> because it doesn't abort due to an issue with `__WinAPI_GetTcpTable`.

if the issue occurs then port will be 0, we can check if the error is general or not found and in case generall trying to set it manually.

Reply:
> whereas your version simply exits without any type of explanation AFAICS.

`_WD_GetFreePort` should be modified to provide `$sMessage`

Reply:
We can also force to return `$iMinPort` on `$_WD_ERROR_NotFound`

Reply:
> _WD_GetFreePort should be modified to provide $sMessage

It looks like the current call to `__WD_Error` is currently missing that parameter. 

> if the issue occurs then port will be 0, we can check if the error is general or not found and in case generall trying to set it manually.

This sounds like what I suggested, but with a check for which error occurred.

> In case $_WD_ERROR_NotFound we can also force to return $iMinPort instead 0

Yes, that is certainly an option.

Reply:
See #457

Reply:
> See #457

Saw.
It will be ok all together.